src/tpm2_dictionarylockout.c

```diff
@@ -0,0 +1,229 @@
+// may be used to endorse or promote products derived from this software without
+// specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <stdbool.h>
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <getopt.h>
+
+#include <sapi/tpm20.h>
+
+#include "log.h"
+#include "main.h"
+#include "options.h"
+#include "string-bytes.h"
+
+#define PASSWORD_MAX (sizeof(((TPM2B_DIGEST *)NULL)->t.buffer))
+
+typedef struct dictionarylockout_ctx dictionarylockout_ctx;
+struct dictionarylockout_ctx {
+    UINT32 max_tries;
+    UINT32 recovery_time;
+    UINT32 lockout_recovery;
+    char lockout_passwd[PASSWORD_MAX];
+    bool clear_lockout;
+    bool setup_parameters;
+    bool use_passwd;
+    TSS2_SYS_CONTEXT *sapi_context;
+};
+
+bool dictionary_lockout_reset_and_parameter_setup(dictionarylockout_ctx *ctx) {
+    
+    TPMS_AUTH_COMMAND sessionData;
+    TPMS_AUTH_RESPONSE sessionDataOut;
+    TSS2_SYS_CMD_AUTHS sessionsData;
+    TSS2_SYS_RSP_AUTHS sessionsDataOut;
+    TPMS_AUTH_COMMAND *sessionDataArray[1];
+    TPMS_AUTH_RESPONSE *sessionDataOutArray[1];
+
+    sessionDataArray[0] = &sessionData;
+    sessionDataOutArray[0] = &sessionDataOut;
+    sessionsDataOut.rspAuths = &sessionDataOutArray[0];
+    sessionsData.cmdAuths = &sessionDataArray[0];
+    sessionsDataOut.rspAuthsCount = 1;
+    sessionData.sessionHandle = TPM_RS_PW;
+    sessionData.nonce.t.size = 0;
+    sessionData.hmac.t.size = 0;
+    *( (UINT8 *)((void *)&sessionData.sessionAttributes ) ) = 0;
+    sessionsData.cmdAuthsCount = 1;
+    sessionsData.cmdAuths[0] = &sessionData;
+
+    if(ctx->use_passwd == true) {
+        sessionData.hmac.t.size = sizeof(sessionData.hmac.t) - 2;
+        if(str2ByteStructure(ctx->lockout_passwd,&sessionData.hmac.t.size,
+                        sessionData.hmac.t.buffer) != 0) {
+            LOG_ERR("Error password could not be used.\n"); 
+        }
+    }
+
+    /*
+    * If setup params and clear lockout are both required, clear lockout should
+    * preceed parameters setup.
+    */
+    UINT32 rval;
+    if(ctx->clear_lockout==true) {
```

Comment:
style: if (ctx->clear_lockout == true) {
...
}

make that change here and throughout.

I prefer, but Ill let that be an optional change, the below:

if (ctx->clear_lockout) {